blockchain/My-Wallet-V3-iOS
# feat(HB): IOS-1319 & IOS-1336 - Error handling in the Exchange

## Objective

⚠️Large PR

This PR includes a few things:
* Significant refactor of how we handle UI updates in `ExchangeCreateViewController`
* Additional classes for having a completion callback that executes an array of `CompletionEvent` types. You can use this on regular UIView animated updates and UIView transition animations. This was required for both as animating a color change of a UILabel is a transition. 
* Additional logic in how we fetch the `mix` and `max` values for trade limits. We now wrap these two requests in an `Observable`. When we receive them we compare them to the user input. **NOTE**: This only reports in fiat currently. 

⚠️This includes some changes from @woudini. Some of those changes have been removed as we learned that `TradeLimits` takes into account the user's balance. That said, not all of it was removed. I'll leave what should and shouldn't be removed up to @woudini.  

Lastly, this code isn't as documented as I would like, and the naming of certain `typealiases` need to be updated. I will do this in a follow up PR. 

## Description

`ExchangeCreatePresenter` now has an `InternalEvent` enum. Currently there's only one type of `.block`. It's a `CompletionEvent`. When a `TransitionUpdateGroup` or `ViewUpdateGroup` completes all of its animations, it calls `finish()`. If it has a completion handler, it calls all the `CompletionEvents`. Here's an example:

```
let completion: ViewUpdateBlock = { [weak self] internalEvents in
            guard let this = self else { return }
            guard let events = internalEvents else { return }
            events.forEach({ this.handle(internalEvent: $0) })
        }
        
        let block = { [weak self] in
            guard let this = self else { return }
            this.setErrorDisappearanceTimer(duration: 2.5)
        }
        
        let group = ViewUpdateGroup(
            preparations: [.secondaryLabel(.hidden)],
            animations: [.errorLabel(.visible)],
            animation: .standard(duration: 0.2),
            completionEvents: [.block(block)],
            completion: completion
        )
        interface?.apply(presentationUpdateGroup: group)
```
We show the `errorLabel` using a `standard` animation with a duration of `0.2`. When it completes it calls the `block` which sets a `Timer` for a duration of `2.5` seconds. When that timer completes, we hid the `errorLabel`:
```
fileprivate func setErrorDisappearanceTimer(duration: TimeInterval) {
        errorDisappearenceTimer?.invalidate()
        errorDisappearenceTimer = Timer(
            fire: Date(timeIntervalSinceNow: duration),
            interval: 0,
            repeats: false,
            block: { [weak self] timer in
                self?.errorDisappearanceTimerFired()
        })
        guard let timer = errorDisappearenceTimer else { return }
        RunLoop.main.add(timer, forMode: .commonModes)
    }
```

## How to Test

1. Build and run
2. Go to the exchange
3. Try to make trades above your min or max limit
4. The primary label should turn red, the primary label should wiggle, and the error label should appear. 

## Screenshot/Design assessment

![simulator screen shot - iphone x - 2018-10-03 at 13 16 00](https://user-images.githubusercontent.com/41585563/46430453-951d6000-c70e-11e8-9950-f2b33fe7a2c3.png)

## Merge Checklist
